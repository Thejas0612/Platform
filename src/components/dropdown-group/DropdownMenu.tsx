import React, { useState } from "react";
import { FormControl, InputLabel, MenuItem, Select, SelectChangeEvent } from "@mui/material";
import KeyboardArrowDownIcon from "@mui/icons-material/KeyboardArrowDown";

const selectStyles = {
  boxShadow: "none",
  ".MuiOutlinedInput-notchedOutline": { borderBottom: 0, borderRight: 0, borderLeft: 0 },
  "&.MuiOutlinedInput-root:hover .MuiOutlinedInput-notchedOutline": {
    border: 0
  },
  "&.MuiOutlinedInput-root.Mui-focused .MuiOutlinedInput-notchedOutline": {
    border: 0
  }
}

export interface Option {
  value: string;
  label: string;
}

interface DropdownProps {
  id: string;
  label: string;
  options: Option[];
  onChange: (id: string, value: string) => void;
}

const DropdownMenu: React.FC<DropdownProps> = ({ id, label, options, onChange }) => {
  const [value, setValue] = useState("");
  const handleChange = (event: SelectChangeEvent<unknown>) => {
    const selectedValue = event.target.value as string;
    setValue(selectedValue);
    onChange(id, selectedValue);
  };
  const isValueSelected = value !== "";
  return (
    <FormControl fullWidth>
      <InputLabel shrink={isValueSelected}>{label}</InputLabel>
      <Select
        onChange={handleChange}
        sx={selectStyles}
        IconComponent={KeyboardArrowDownIcon}
        value={value}
        label={label}
      >
        {options.map((option, index) => (
          <MenuItem key={index} value={option.value}>
            {option.label}
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
};

export default DropdownMenu;
