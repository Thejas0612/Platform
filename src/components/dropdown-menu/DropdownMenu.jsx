import React from "react";
import FormControl from "@mui/material/FormControl";
import Select from "@mui/material/Select";
import MenuItem from "@mui/material/MenuItem";
import InfoIcon from "@mui/icons-material/Info";
import IconButton from "@mui/material/IconButton";
import Tooltip from "@mui/material/Tooltip";
import { Grid } from "@mui/material";
import axios from "axios";
import { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";

import { updateRightSection } from "../../../../EMR-ENT-MSOL-PA-PLATFORM-CLIENT/src/redux/reducers/initialBuDataSlice";

function DropdownMenu(schema) {
    
  // const dataOptions = useSelector((state) => state.initialBuData?.rightSection[0]?.componentProps?.schema[1]?.fields[4]?.options);
  const dataOptions = useSelector((state) => state.initialBuData?.rightSection);
  // apiData
  const updatedSchema = (apiData, dataOptions,target, key='test', value='test') => {
    const schema = dataOptions[0]?.componentProps?.schema;


    const updatedSchema = schema.map((option, i) => {
      if (i === 1) {
        return {
          ...option,
          fields: option.fields.map((field) =>
            field.name === target ? { ...field, options: apiData, isApiOnEvent: { ...field.isApiOnEvent, apiInfo: { ...field.isApiOnEvent.apiInfo, payload: { [key]: value } } } } : field
          )
        };
      } else {
        return option;
      }
    });

    //   const updatedSchema = schema.map((option, i) => {
    //     if (i === 1) {

    //       return {
    //         ...option,
    //         fields: option.fields.map((field, j) =>
    //           j === 4 ? { ...field, dataSourceUrl: 'https://webapp-z-autosol-msolst-n-001.azurewebsites.net/api/temp/form-factor?communication_protocol=4_20_MA_WITH_HART_PROTOCOL' } : field
    //         )
    //       };
    //     } else {
    //       return option;
    //     }
    //   });
    //   console.log("updated_schema",updatedSchema)

    const updatedDataOptions = [
      {
        ...dataOptions[0],
        componentProps: {
          ...dataOptions[0].componentProps,
          schema: updatedSchema
        }
      }
    ];

    return { updatedDataOptions };
  };

  const dispatch = useDispatch();
  const [selectValue, setSelectValue] = React.useState("");
  const [data, setData] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(null);



  useEffect(() => {
    const fetchData = async (url) => {
      try {
        const data = await axios.get(url);
        const apiData = data.data
        const result = updatedSchema(apiData, dataOptions,'communicationProtocol');
        console.log('result..........',result)
        dispatch(updateRightSection(result.updatedDataOptions));
        setIsLoading(false);

        // setData(response.data);
      } catch (error) {
        setError(error);
      } finally {
        setIsLoading(false);
      }
    };
    if (schema.dataSourceUrl) {
      fetchData(schema.dataSourceUrl);
    }
  }, []);

  const [defaultValue, setDefaultValue] = React.useState("");
  const [dropValue, SetDropValue] = React.useState({communicationProtocol:''})
  const [dropValue1, SetDropValue1] = React.useState({formfactor:''})
  

  const handleChange = async (e) => {
    const value = e.target.value;
    const name = e.target.name;
    console.log('love',schema)
    console.log("get_me", e);
    setSelectValue(value);
    if (schema.name === name) {
      if (name === "communicationProtocol") {
        const target = schema.isApiOnEvent.targetUiElement
        const getUrl = schema.isApiOnEvent.apiInfo.url;
        const data = await axios.get(getUrl + value);
        const apiData = data.data;
        const result = updatedSchema(apiData, dataOptions,target,'communicationProtocol',value );
        console.log('uuuuuuuuuuuu',result)
        dispatch(updateRightSection(result.updatedDataOptions));
        setIsLoading(false);
        setDefaultValue(e.target.value);
        SetDropValue({communicationProtocol: e.target.value})

      }else if (name === "formfactor"){
        const target = schema.isApiOnEvent.targetUiElement
        const baseUrl = schema.isApiOnEvent.apiInfo.url;
        const savedCommunicationProtocol = schema.isApiOnEvent.apiInfo.payload.communicationProtocol
        const savedFormFactor = value;
        console.log(typeof(value),'aaaaaaaaaaaaaa', value)
        const url = `${baseUrl}form_factor=${savedFormFactor}&communication_protocol=${savedCommunicationProtocol}`;
        const data = await axios.get(url);
        const apiData = data.data;
        const result = updatedSchema(apiData, dataOptions,target,'formFactor',value );
        console.log('uuuuuuuuuuuu',result)
        dispatch(updateRightSection(result.updatedDataOptions));
        setIsLoading(false);
        setDefaultValue(e.target.value)
        SetDropValue1({formfactor: e.target.value})
        
      }else if (name === "numberofinputs"){
        const target = schema.isApiOnEvent.targetUiElement
        const baseUrl = schema.isApiOnEvent.apiInfo.url;
        // const savedCommunicationProtocol = schema.isApiOnEvent.apiInfo.payload.communicationProtocol
        const savedFormFactor = schema.isApiOnEvent.apiInfo.payload.formFactor
        const savedNumberOfInputs = value;
        console.log('jSSSS',savedFormFactor, savedNumberOfInputs, commu)
        // console.log(typeof(value),'aaaaaaaaaaaaaa', value)
        // const url = `${baseUrl}form_factor=${savedFormFactor}&communication_protocol=${savedCommunicationProtocol}`;
        // const data = await axios.get(url);
        // const apiData = data.data;
        // const result = updatedSchema(apiData, dataOptions,target,'formFactor',value );
        // console.log('uuuuuuuuuuuu',result)
        // dispatch(updateRightSection(result.updatedDataOptions));
        // setIsLoading(false);
        // setDefaultValue(e.target.value)
        // SetDropValue1({formfactor: e.target.value})
        
      }
    }
  };

  //     if (schema.isApiOnEvent?.isApiCall && schema.isApiOnEvent.apiInfo?.url && value) {
  //         try {

  //         const data = await axios.get(schema.isApiOnEvent.apiInfo.url + value);
  //         const apiData =data.data
  //         console.log('3',apiData);
  //         const result = updatedSchema(apiData, dataOptions);
  //         console.log('result.....................',result)
  //         dispatch(updateRightSection(result.updatedDataOptions))
  //         setIsLoading(false)
  //         // setData(apiData.data);

  //     } catch (error) {
  //         setError(error);
  //     }
  //         //const response = fetchData();

  //     }
  //     setDefaultValue(e.target.value)
  // };
  return (
    <div>
      <FormControl
        sx={{ m: 1, mt: 3, margin: "1rem", width: "100%", display: "flex" }}
        size="small"
        color="success"
      >
        <Grid container rowSpacing={1} columnSpacing={{ xs: 1, sm: 2, md: 3 }}>
          <Grid item xs={12}>
            <div style={{ minHeight: "2rem" }}>
              <label>{schema.label}</label>
              {schema.tooltipEnable ? (
                <Tooltip title={schema.tooltipMessage} placement="top" arrow>
                  <IconButton sx={{ padding: 0, paddingLeft: ".5rem" }}>
                    <InfoIcon sx={{ color: "#00573d" }} fontSize="small" />
                  </IconButton>
                </Tooltip>
              ) : (
                <></>
              )}
            </div>
            <Select
              sx={{ width: "100%" }}
              value={defaultValue}
              name={schema.name}
              onChange={handleChange}
            >
              {/* {
                isLoading ? (
                  <option>loading...</option>
                ) : error ? (
                  <option>Error: {error.message}</option>
                ) : (
                  data.map((option) => (
                    <MenuItem key={option.id} value={option.value} name="Sourabh">
                      {option.label}
                    </MenuItem>
                  ))
                )
                // (schema.options.length > 0 &&
                //     schema.options.map((opt, i) => (
                //         <MenuItem key={i} value={opt.label} onChange={handleChange}>
                //             {opt.label}
                //         </MenuItem>
                //     )))
              } */}
              {schema.options.length > 0 &&
                schema.options.map((opt, i) => (
                  <MenuItem key={i} value={opt.value}>
                    {opt.label}
                  </MenuItem>
                ))}
            </Select>
          </Grid>
        </Grid>
      </FormControl>
      {schema.nextLine ? <br></br> : ""}
    </div>
  );
}

export default DropdownMenu;
