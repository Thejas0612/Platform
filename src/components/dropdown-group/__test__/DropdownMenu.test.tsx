import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import DropdownMenu, { Option } from '../DropdownMenu';

describe('DropdownMenu', () => {
  const options: Option[] = [
    { value: 'yes', label: 'Yes' },
    { value: 'no', label: 'No' }
  ];

  it('renders DropdownMenu correctly', async() => {
    const onChangeMock = jest.fn();
     render(
      <DropdownMenu id="dropdown" label="Dropdown" options={options} onChange={onChangeMock} />
    );
    const label = await screen.findByRole('combobox')
  });


it('renders DropdownMenu correctly with options', async () => {
  const onChangeMock = jest.fn();
  const {container} = render(
      <DropdownMenu id="dropdown" label="Dropdown" options={options} onChange={onChangeMock} />
    );
    const comboBox = screen.getByRole('combobox');
    fireEvent.mouseDown(comboBox);

    options.forEach(option => {
      const optionElement = screen.getByText(option.label);
      expect(optionElement).toBeInTheDocument();
    });
    expect(container).toMatchSnapshot()
  });
  
});