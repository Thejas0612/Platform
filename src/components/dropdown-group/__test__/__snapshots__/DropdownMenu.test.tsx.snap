// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenu renders DropdownMenu correctly with options 1`] = `
<div
  aria-hidden="true"
>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
    >
      Dropdown
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl  css-1kd47wj-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
    >
      <div
        aria-controls="mui-2"
        aria-expanded="true"
        aria-haspopup="listbox"
        class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        role="combobox"
        tabindex="0"
      >
        <span
          class="notranslate"
        >
          â€‹
        </span>
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined MuiSelect-iconOpen css-bpeome-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="KeyboardArrowDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-14lo706"
        >
          <span>
            Dropdown
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
